File: src/main/java/ru/albemuth/analysis/population/Human.java
package ru.albemuth.analysis.population;

/**
 * @author VKornyshev
 * Класс, моделирующий человека
 */
public abstract class Human {

    public enum Sex {
        MALE, FEMALE
    }

    protected int ageLimit;//предельный возраст, по достижении которого человек умирает
    protected int age;//возраст
    protected double firstLaborDeathP;//вероятность смерти первыми родами

    public Human(int ageLimit, double firstLaborDeathP) {
        this.ageLimit = ageLimit;
        this.age = 0;
        this.firstLaborDeathP = firstLaborDeathP;
    }

    public abstract Sex getSex();

    public int getAge() {
        return age;
    }

    public void setAge(int age) {
        this.age = age;
    }

    public void incAge() {
        age++;
    }

    public boolean isDead() {
        return age >= ageLimit;
    }

    public double getFirstLaborDeathP() {
        return firstLaborDeathP;
    }

}

File: src/main/scala/ru/albemuth/analysis/population/v3/Human.scala
package ru.albemuth.analysis.population.v3

import scala.math.random

/**
 * @author vovan
 * Класс, моделирующий человека
 */
object Human {

    object Sex extends Enumeration {
        type Sex = Value

        val MALE = Value("male")
        val FEMALE = Value("female")
    }

    def firstYearDeathP(d: Double): Double = {
        import scala.math.sqrt
        import scala.math.Pi
        import scala.math.exp
        val m = 0.35
        val s = 0.12930984560277864
        val c = 0.32639851212114895

        1 - Population.p(c / (s * sqrt(2 * Pi)) * exp(- (d - m) * (d - m) / (2 * s * s)))
    }

}

abstract class Human(val cranialDiameter: Double, var age: Int, val ageLimit: Int) {

    import Human._
    import Human.Sex._
    def sex: Sex
    private var dead = false

    def incAge(): Int = {
        age += 1
        dead = age == 1 && random < firstYearDeathP(cranialDiameter) || age >= ageLimit
        age
    }

    def isDead: Boolean = dead

}
